### PR TITLE
Feature/#232 validation method jwt

### PR DESCRIPTION
**Pull request by nataschaberg**
Implements #232

### CODE REVIEW

File: ts/utils/helper.ts [hunk 1]
```diff
@@ -1,5 +1,31 @@
...
+
+export function getIssuerPublicKey(keyId: string, ddo: DidDocument): Buffer {
+  const relevantPubKey = ddo.getPublicKeySections().map((keySection) => {
+    if (keySection.getIdentifier() === keyId) { return keySection.getPublicKeyHex() }
+  })
+
+  return Buffer.from(relevantPubKey[0], 'hex')
+}
+
+export function handleValidationStatus(status: boolean, key: string) {
```
Comment (Exulansis) [inline ts/utils/helper.ts]: Why switch statement, perhaps:

```typescript
export function handleValidationStatus(success: boolean, invalidKey: string) {
  if (!success) {
    throw new Error(ErrorKeys[key] || 'Unknown Error key') 
  }
}
```

File: ts/utils/helper.ts [hunk 2]
```diff
@@ -1,5 +1,31 @@
...
+  switch (status) {
+    case false:
+      throw new Error(ErrorKeys[key] || 'Unknown Error key') 
+    default:
+      return
+  }
+}
+
+export enum ErrorKeys {
```
Comment (Exulansis) [inline ts/utils/helper.ts]: This could also be a simple object, `enums` are usefull when we want to reuse this in other files

Comment (Exulansis) [reply ts/utils/helper.ts]: We also don't need to export this, since it's never used outside of this file.

File: ts/utils/helper.ts [hunk 3]
```diff
@@ -1,5 +1,31 @@
+import { DidDocument } from "../identity/didDocument/didDocument"
+
 // Helper function for data preparation
 
 export function keyIdToDid(keyId: string): string {
   return keyId.substring(0, keyId.indexOf('#'))
+}
+
+export function getIssuerPublicKey(keyId: string, ddo: DidDocument): Buffer {
+  const relevantPubKey = ddo.getPublicKeySections().map((keySection) => {
```
Comment (Exulansis) [inline ts/utils/helper.ts]: We can use `ddo.getPublicKeySections().find(section => section.getIdentifier() === keyId)` for brevity.

File: ts/interactionTokens/JSONWebToken.ts [hunk 4]
```diff
@@ -153,10 +154,7 @@ export class JSONWebToken<T extends JWTEncodable> implements IDigestable {
 
   public static decode<T extends JWTEncodable>(jwt: string): JSONWebToken<T> {
     const interactionToken = JSONWebToken.fromJSON(decodeToken(jwt))
-
-    if (interactionToken.getExpirationTime() <= Date.now()) {
-      throw new Error('Token expired')
-    }
+    handleValidationStatus((interactionToken.getExpirationTime() - 1 > Date.now()), 'exp')
```
Comment (Exulansis) [inline ts/interactionTokens/JSONWebToken.ts]: Do we need the `- 1`?

File: ts/identityWallet/identityWallet.ts [hunk 5]
```diff
@@ -170,6 +174,16 @@ export class IdentityWallet {
     return jwt
   }
 
+  public async validateJWT<T extends JWTEncodable>(jwtReceived: JSONWebToken<T>, jwtSend?: JSONWebToken<T>, customRegistry?: JolocomRegistry): Promise<void> {
+    const registry = customRegistry || createJolocomRegistry()
+    const ddo = await registry.resolve(keyIdToDid(jwtReceived.getIssuer()))
```
Comment (Exulansis) [inline ts/identityWallet/identityWallet.ts]: Perhaps rename this to `identity`, it is not just the ddo.

File: ts/identityWallet/identityWallet.ts [hunk 6]
```diff
@@ -170,6 +174,16 @@ export class IdentityWallet {
...
+    const pubKey  = getIssuerPublicKey(jwtReceived.getIssuer(), ddo.getDidDocument())
+ 
+    handleValidationStatus(await this.vaultedKeyProvider.verifyDigestable(pubKey, jwtReceived), 'sig')
+    jwtSend && handleValidationStatus(jwtReceived.getAudience() === this.identity.getDid(), 'aud')
```
Comment (Exulansis) [inline ts/identityWallet/identityWallet.ts]: perhaps rename to something more expressive, eg `requestInteractionToken`, `receivedInteractionToken` or something along those lines.

File: tests/identityWallet/validateJWT.test.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      }
+    })
+  })
+
+  after(() => {
+    sandbox.restore()
+  })
+
+  it('Should sucessfully perform necessary validation steps on received jwt', () => {
+    expect(async () => await iw.validateJWT(JSONWebToken.fromJSON(validSignedCredReqJWT))).not.to.throw()
```
Comment (Exulansis) [inline tests/identityWallet/validateJWT.test.ts]: Does this test fail?

File: tests/identityWallet/identityWallet.test.ts [hunk 8]
```diff
@@ -14,21 +14,22 @@ import { credentialResponseJSON } from '../data/interactionTokens/credentialResp
...
 
 chai.use(sinonChai)
 const expect = chai.expect
 
 /* Saves some space during stubbing, helper functions */
 
 const stubbedKeyProvider = {
-  signDigestable: sinon.stub().callsFake(attributes => Buffer.from(signedSimpleCredReqJWT.signature, 'hex')),
+  signDigestable: sinon.stub().callsFake(attributes => Buffer.from(validSignedCredReqJWT.signature, 'hex')),
+  verifyDigestable: sinon.stub().callsFake(attrs => true)
```
Comment (Exulansis) [inline tests/identityWallet/identityWallet.test.ts]: can be replaced with

```typescript
sinon.stub().returns(true)
```

File: tests/identityWallet/identityWallet.test.ts [hunk 9]
```diff
@@ -14,21 +14,22 @@ import { credentialResponseJSON } from '../data/interactionTokens/credentialResp
...
+import { keyIdToDid } from '../../ts/utils/helper'
 
 chai.use(sinonChai)
 const expect = chai.expect
 
 /* Saves some space during stubbing, helper functions */
 
 const stubbedKeyProvider = {
-  signDigestable: sinon.stub().callsFake(attributes => Buffer.from(signedSimpleCredReqJWT.signature, 'hex')),
+  signDigestable: sinon.stub().callsFake(attributes => Buffer.from(validSignedCredReqJWT.signature, 'hex')),
```
Comment (Exulansis) [inline tests/identityWallet/identityWallet.test.ts]: can be replaced with

```typescript
sinon.stub().returns(Buffer.from(...))
```

File: tests/identityWallet/validateJWT.test.ts (hunk 7 above)
Comment (nataschaberg) [reply tests/identityWallet/validateJWT.test.ts]: Fails when signature is invalid which is tested explicitly in the case below this one.

**Pull request merged**